serve/mlc_serve/engine/engine_common.py

```diff
@@ -0,0 +1,336 @@
+"""
+Common utilites for engine classes.
+"""
+
+import time
+from typing import Tuple, Deque, Dict, Optional, Union
+from collections import deque
+from threading import Condition, Lock
+
+import structlog
+
+from .base import (
+    Request,
+    RequestId,
+    RequestState,
+    GenerationSequence,
+    SequenceId,
+    StoppingCriteria,
+    check_stopping_sequences,
+)
+from .metrics import PrometheusMetrics
+from .metrics_labels import NUM_CACHE_EVICTONS
+from .model_module import (
+    DecodeRequest,
+    PrefillRequest,
+    Tokenizer,
+    ConversationTemplate,
+    KVCacheManager,
+    ModelModule,
+    TextGenerator,
+    Tokenizer as TokenizerP,
+)
+from ..model.base import ModelArtifactConfig
+
+LOG = structlog.stdlib.get_logger(__name__)
+
+
+def get_new_request_state(
+    request: Request, conversation_template: ConversationTemplate, tokenizer: Tokenizer
+) -> RequestState:
+    if request.debug_options.prompt is not None:
+        prompt = request.debug_options.prompt
+    else:
+        prompt = conversation_template.apply(request.messages)
+
+    prompt_tokens = tokenizer.encode(prompt)
+
+    validation_err = None
+    if request.validate_tokens is not None:
+        validation_err = request.validate_tokens(request, prompt_tokens)
+
+    gen_seqs = [
+        GenerationSequence(
+            seq_id=SequenceId(request.request_id, i),
+            generated_token_ids=[],
+            next_start_position=0,
+            output_text="",
+        )
+        for i in range(request.num_sequences)
+    ]
+
+    return RequestState(
+        request_id=request.request_id,
+        prompt_token_ids=prompt_tokens,
+        generation_sequences=gen_seqs,
+        sampling_params=request.sampling_params,
+        stopping_criteria=request.stopping_criteria,
+        debug_options=request.debug_options,
+        validation_err=validation_err,
+        arrival_timestamp=time.time(),
+    )
+
+
+def decode_last_output(
+    prompt_tokens: list[int],
+    generation_sequence: GenerationSequence,
+    tokenizer: Tokenizer,
+) -> str:
+    if len(generation_sequence.output_text):
+        prefix_idx = max(0, generation_sequence.next_start_position - 6)
+    else:
+        prefix_idx = generation_sequence.next_start_position
+
+    # TODO(masahi): Figure out a way to remove this concat
+    token_ids = prompt_tokens + generation_sequence.generated_token_ids
```

Comment:
This list concat is probably causing perf regression. Improving this is left for future work.

Reply:
PR https://github.com/octoml/mlc-llm/pull/102 changes this part. Although it adds more concats and I also agree with @elvin-n, it would be good to collect data and confirm.